py_trees/decorators.py

```diff
@@ -920,3 +922,74 @@ def update(self) -> common.Status:
             the behaviour's new status :class:`~py_trees.common.Status`
         """
         return self.decorated.status
+
+
+class ForEach(Decorator):
+    """
+    Run the child behavior for each item in an iterable.
+
+    On initialization, the iterable is loaded from the blackboard and an
+    iterator is created. Every time the child succeeds, we advance to
+    the next item. We keep running until the iterable is exhausted.
+    """
+
+    def __init__(
+        self, name: str, child: behaviour.Behaviour, source_key: str, target_key: str
+    ):
+        """
+        Initialise the ForEach decorator.
+
+        Args:
+            name (:obj:`str`): name of the behaviour
+            child (:obj:`Behaviour`): the child behaviour to decorate
+            source_key (:obj:`str`): blackboard key to read the input iterable
+            target_key (:obj:`str`): blackboard key to set for each iteration
+        """
+        super().__init__(name=name, child=child)
+        self.source_key = source_key
+        self.target_key = target_key
+        self.blackboard = blackboard.Client(name=name)
+        self.blackboard.register_key(key=self.source_key, access=common.Access.READ)
+        self.blackboard.register_key(key=self.target_key, access=common.Access.WRITE)
+        self._iterator: typing.Optional[typing.Iterator] = None
```

Comment:
I think `typing.Optional` is falling out of favor for this

```suggestion
        self._iterator: typing.Iterator | None = None